sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/DoNotCheckZeroSizeCollection.cs

```diff
@@ -0,0 +1,144 @@
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+using System;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class DoNotCheckZeroSizeCollection : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S3981";
+        private const string MessageFormat = "The {0} of '{1}' is always '>=0', so fix this test to get the real expected behavior.";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var gteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, gteNode.Left, gteNode.Right);
+                }, SyntaxKind.GreaterThanOrEqualExpression);
+
+            context.RegisterSyntaxNodeActionInNonGenerated(c =>
+                {
+                    var lteNode = (BinaryExpressionSyntax)c.Node;
+                    CheckCondition(c, lteNode.Right, lteNode.Left);
+                }, SyntaxKind.LessThanOrEqualExpression);
+        }
+
+        private static void CheckCondition(SyntaxNodeAnalysisContext context, ExpressionSyntax expressionValueNode,
+            ExpressionSyntax constValueNode)
+        {
+            if (!IsConstantZero(context, constValueNode))
+            {
+                return;
+            }
+
+            ISymbol symbol = GetSymbol(context, expressionValueNode);
+
+            string symbolType = null;
+            if (IsArrayLengthProperty(symbol))
+            {
+                symbolType = "Array";
+            }
+            else if (IsEnumerableCountMethod(symbol))
+            {
+                symbolType = "IEnumerable<T>";
+            }
+            else if (IsCollectionProperty(symbol))
+            {
+                symbolType = "ICollection";
+            }
+
+            if (symbolType == null)
+            {
+                return;
+            }
+
+            context.ReportDiagnostic(Diagnostic.Create(rule, context.Node.GetLocation(),
+                symbol.Name.ToLower(), symbolType));
+        }
+
+        private static ISymbol GetSymbol(SyntaxNodeAnalysisContext context, ExpressionSyntax expressionValueNode)
+        {
+            while (true)
```

Comment:
Why should it be avoided? It may look a bit weird, but the advantage of that approach is that the variable scope is narrowed down to inside the while loop. The RemoveParentheses puts the condition in the while block and it looks like that:

```
public static ExpressionSyntax RemoveParentheses(this ExpressionSyntax expression)
{
    var currentExpression = expression;
    var parentheses = expression as ParenthesizedExpressionSyntax;
    while (parentheses != null)
    {
        currentExpression = parentheses.Expression;
        parentheses = currentExpression as ParenthesizedExpressionSyntax;
    }
    return currentExpression;
}
```

Rewritten to while(true), it is IMHO much more readable: 

```
public static ExpressionSyntax RemoveParentheses(this ExpressionSyntax expression)
{
    while (true)
    {
        var parentheses = expression as ParenthesizedExpressionSyntax;
        if (parentheses == null)
        {
            break;
        }
        expression = parentheses;
    }
    return expression;
}
```

The risk of going into infinite loop is exactly the same in both cases...